### PR TITLE
Generate warning suppressions in an XML format 

### PR DESCRIPTION
**Pull request by mateoatr**
See #1380. This changes the existing `--generate-warning-suppressions` option to receive a string argument (`cs`/`xml`) which tells the suppression writer in which format should the warning suppressions be emitted, either as a C# source file (this is the current behavior) or as an XML file that follows the linker's annotation format.

/cc @layomia

### CODE REVIEW

File: src/linker/Linker/WarningSuppressionWriter.cs [hunk 1]
```diff
@@ -2,19 +2,30 @@
 using System.IO;
 using System.Linq;
 using System.Text;
+using System.Xml;
+using System.Xml.Linq;
 using Mono.Cecil;
 
 namespace Mono.Linker
 {
+	public enum WarningSuppressionWriterFileOutputKind
```
Comment (marek-safar) [inline src/linker/Linker/WarningSuppressionWriter.cs:11]: Why is it not nested enum inside `WarningSuppressionWriter` ?

File: src/linker/Linker/WarningSuppressionWriter.cs [hunk 2]
```diff
@@ -2,19 +2,30 @@
...
+	{
+		Cs,
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/linker/Linker/WarningSuppressionWriter.cs:13]: I'd also call it CSharp

File: src/linker/Linker/WarningSuppressionWriter.cs [hunk 3]
```diff
@@ -31,46 +42,87 @@ public void AddWarning (int code, IMemberDefinition memberDefinition)
...
+				if (lhs == rhs)
+					return a.Code.CompareTo (b.Code);
+
+				return string.CompareOrdinal (lhs, rhs);
+			});
+
+			return listOfWarnings;
+		}
+
+		string GetWarningSuppressionScopeString (IMemberDefinition member)
```
Comment (marek-safar) [inline src/linker/Linker/WarningSuppressionWriter.cs:113]: Should be static

File: src/linker/Linker/WarningSuppressionWriter.cs [hunk 4]
```diff
@@ -31,46 +42,87 @@ public void AddWarning (int code, IMemberDefinition memberDefinition)
...
+
+				sw.Write (sb.ToString ());
 			}
 		}
+
+		List<(int Code, IMemberDefinition Member)> GetListOfWarnings (AssemblyNameDefinition assemblyName)
+		{
+			List<(int Code, IMemberDefinition Member)> listOfWarnings = _warnings[assemblyName].ToList ();
+			listOfWarnings.Sort ((a, b) => {
+				string lhs = a.Member is MethodReference lhsMethod ? lhsMethod.GetDisplayName () : a.Member.FullName;
```
Comment (marek-safar) [inline src/linker/Linker/WarningSuppressionWriter.cs:102]: Will this logic work for members in nested types? I guess Cecil FullName won't compare well with GetDisplayName

File: src/linker/Linker/Driver.cs [hunk 5]
```diff
@@ -430,8 +430,10 @@ protected int SetupContext (ILogger customLogger = null)
 						if (!GetStringParam (token, l => generateWarningSuppressionsArgument = l))
 							return -1;
 
-						if (!GetWarningSuppressionWriterFileOutputKind (generateWarningSuppressionsArgument, out var fileOutputKind))
+						if (!GetWarningSuppressionWriterFileOutputKind (generateWarningSuppressionsArgument, out var fileOutputKind)) {
+							context.LogError ($"Invalid value '{generateWarningSuppressionsArgument}'specified for '--generate-warning-suppressions' option", 2061);
```
Comment (vitek-karas) [inline src/linker/Linker/Driver.cs:434]:
```suggestion
							context.LogError ($"Invalid value '{generateWarningSuppressionsArgument}' specified for '--generate-warning-suppressions' option", 2061);
```

File: src/linker/Linker/Driver.cs [hunk 6]
```diff
@@ -938,6 +946,22 @@ static bool GetOptimizationName (string text, out CodeOptimizations optimization
...
+			switch (text.ToLowerInvariant ()) {
+			case "cs":
+				fileOutputKind = WarningSuppressionWriterFileOutputKind.Cs;
+				return true;
+			case "xml":
+				fileOutputKind = WarningSuppressionWriterFileOutputKind.Xml;
+				return true;
+			}
+
+			Console.WriteLine ($"Invalid warning suppressions file format '{text}'");
```
[the remark this answers is not in the document]
Comment (mateoatr) [reply src/linker/Linker/Driver.cs]: You're right, following https://github.com/mono/linker/pull/1364/ this should  be an error.

File: src/linker/Linker/WarningSuppressionWriter.cs (hunk 4 above)
Comment (mateoatr) [reply src/linker/Linker/WarningSuppressionWriter.cs:96]: I think I don't understand why would there be a problem with members in nested types. We only use this strings for sorting the list.

Comment (marek-safar) [reply src/linker/Linker/WarningSuppressionWriter.cs:96]: I believe GetDisplayName uses `.` as separator whereas FullName uses `+` as separator

File: src/linker/Linker/WarningSuppressionWriter.cs [hunk 7]
```diff
@@ -31,46 +36,93 @@ public void AddWarning (int code, IMemberDefinition memberDefinition)
...
-
-						DocumentationSignatureGenerator.Instance.VisitMember (warningOrigin, sb);
-						sb.AppendLine ("\")]");
-					}
-
-					sw.Write (sb.ToString ());
+				if (_fileOutputKind == FileOutputKind.Xml)
+					OutputSuppressionsXmlFormat (assemblyName);
+
+				OutputSuppressionsCsFormat (assemblyName);
```
Comment (layomia) [inline src/linker/Linker/WarningSuppressionWriter.cs:42]: Should this be in an `else` branch?

File: src/linker/Linker/WarningSuppressionWriter.cs (hunk 4 above)
Comment (sbomer) [reply src/linker/Linker/WarningSuppressionWriter.cs:99]: I think it'd be good to use the signature to sort, since that's what we're outputing.

**Pull request merged**